src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,640 @@
+            f = _sub_rest(f, mul_term(qi, left, g[match_index], right), 1)
+        else
+            push!(rcoeffs, f.coeffs[1])
+            push!(rexps, f.exps[1])
+            f = FreeAssAlgElem{T}(R, f.coeffs[2:end], f.exps[2:end], length(f) - 1)
+        end
+    end
+    return FreeAssAlgElem{T}(R, rcoeffs, rexps, length(rcoeffs))
+end
+
+# normal form with leftmost word divisions
+function normal_form(
+    f::FreeAssAlgElem{T},
+    g::Vector{FreeAssAlgElem{T}},
+) where {T<:FieldElement}
+    R = parent(f)
+    s = length(g)
+    rcoeffs = T[]
+    rexps = Monomial[]
+    while length(f) > 0
+        i = 1
+        @label again
+        ok, ml, mr = word_divides_leftmost(f.exps[1], g[i].exps[1])
+        if !ok && i < s
+            i += 1
+            @goto again
+        end
+        if ok
+            qi = divexact(f.coeffs[1], g[i].coeffs[1])
+            f = _sub_rest(f, mul_term(qi, ml, g[i], mr), 1) # enforce lt cancelation
+        else
+            push!(rcoeffs, f.coeffs[1])
+            push!(rexps, f.exps[1])
+            f = FreeAssAlgElem{T}(R, f.coeffs[2:end], f.exps[2:end], length(f) - 1)
+        end
+    end
+    r = FreeAssAlgElem{T}(R, rcoeffs, rexps, length(rcoeffs))
+    return r
+end
+
+# weak normal form with leftmost word divisions
+function normal_form_weak(
+    f::FreeAssAlgElem{T},
+    g::Vector{FreeAssAlgElem{T}},
+) where {T<:FieldElement}
+    R = parent(f)
+    s = length(g)
+    while length(f) > 0
+        i = 1
+        @label again
+        ok, ml, mr = word_divides_leftmost(f.exps[1], g[i].exps[1])
+        if !ok && i < s
+            i += 1
+            @goto again
+        end
+        if ok
+            qi = divexact(f.coeffs[1], g[i].coeffs[1])
+            f = _sub_rest(f, mul_term(qi, ml, g[i], mr), 1) # enforce lt cancelation
+        else
+            break
+        end
+    end
+    return f
+end
+
+@doc raw"""
+    interreduce!(g::Vector{FreeAssAlgElem{T}})
+
+Interreduce a given Groebner basis with itself, i.e. compute the normal form of each
+element of `g` with respect to the rest of the elements and discard elements with
+normal form $0$ and duplicates.
+""" 
+function interreduce!(g::Vector{FreeAssAlgElem{T}}) where {T}
+    i = 1
+    while length(g) > 1 && length(g) >= i
+        aut = AhoCorasickAutomaton([g_j.exps[1] for g_j in g[1:end .!= i]])
+        r = normal_form(g[i], g[1:end .!= i], aut)
+        if iszero(r)
+            deleteat!(g, i)
+        elseif g[i] != r
+            g[i] = r
+            i = 1
+        else
+            i += 1
+        end
+    end
+    return g
+end
+
+## checks whether there is an overlap between a and b at position i of b
+#  such that b[i:length(b)] = a[1:length(b)-i]
+function check_left_overlap(a::Monomial, b::Monomial, i::Int)
+    if length(b) - i >= length(a)
+        return false # this is a not a left overlap but might be a center overlap
+    end
+    for j in 0:(length(b) - i)
+        if b[i + j] != a[j + 1]
```

Comment:
Just to point out an example where `Monomial = Vector{UInt8}` can help: this function could essentially be turned into a call to `memcmp`, which is *fast*.

Actually this could always be turned into a `memcmp` call, but with `UInt8` instead of `Int`, it becomes roughly 8 times faster.